### DISCUSSION TITLE
Nested Teams

### DISCUSSION THREAD

**Discussion by isaiahdahl** [spatie/laravel-permission#2566] [Ideas]
We are building a feature that has nested teams logic, where in principle, child teams inherit the permissions of the parent team. But it doesn't automatically add a person to a parent team. Only the permissions. 

_Use belongs to an org, 
Org has teams,
Team can have child-teams 
child-teams can have child-teams_

What we want to build mirrors this https://github.blog/2017-06-13-nested-teams-add-depth-to-your-team-structure/

Would it be possible to add this as a feature to this library? In my head I'm thinking we add a nullable `parent_team_id` column where `team_id` columns have been added and then adjust the can() ...etc logic to follow the rule that child teams inherit their parent permissions. 

Thoughts?

**Comment by drbyte** [comment]
You raise an interesting idea.

I often find myself imagining a similar hierarchy when brainstorming an app.
And then when I get into implementation I frequently re-discover that Roles themselves are a "team", and the combination of roles assigned to different people often provides sufficient accommodation of "actual" needs in many apps.

And of course, there's still the legitimate scenario where "teams" is truly needed. That ranges from "departments" and some forms of multitenancy to online classroom concepts.
In fact the Teams feature in this package solved a lot of teams and somewhat-nested teams requirements. Thanks @erikn69 for the bulk of the coding of it.

There are probably some scenarios where truly nested teams would be needed.
In such a situation, I find myself wanting to know more about your proposed app structure, and how you envision implementing "teams" (without roles/permissions, just the "rest" of the "teams" stuff), because perhaps that will expose whether the roles/permissions part truly needs its own nesting, or whether your Teams/whatever model can leverage what's already available.

So, all that to say: let's continue the discussion. 
Can you elaborate more on how you intend to architect your project? Especially outside of the roles/permissions part, although I realize sometimes those are quite intertwined.

**Reply by isaiahdahl** [reply]
Thanks for your reply, your statement about "re-discovering" that Roles themselves are a "Team" has spun me into a way of thinking I hadn't considered. Will take some time with that and write back here an overview of where we land.
